src/ucp/stream/stream_recv.c

```diff
@@ -30,110 +62,311 @@ UCS_PROFILE_FUNC(ucs_status_ptr_t, ucp_stream_recv_data_nb,
+    req->recv.stream.count = 0;
+
+    if (ucs_unlikely(UCP_DT_IS_IOV(datatype))) {
+        req->recv.state.dt.iov.iov_offset    = 0;
+        req->recv.state.dt.iov.iovcnt_offset = 0;
+        req->recv.state.dt.iov.iovcnt        = count;
+        req->recv.state.dt.iov.memh          = UCT_MEM_HANDLE_NULL;
+    }
+
+    req->recv.buffer   = buffer;
+    req->recv.datatype = datatype;
+    req->recv.length   = ucp_dt_length(datatype, count, buffer,
+                                       &req->recv.state);
+
+    return req;
+}
+
+static UCS_F_ALWAYS_INLINE ucs_status_ptr_t
+ucp_stream_recv_request_put(ucp_request_t *req, size_t *length,
+                            ucs_status_t status)
+{
+    *length = req->recv.state.offset;
+    ucp_request_put(req);
+    return UCS_STATUS_PTR(status);
+}
+
+UCS_PROFILE_FUNC(ucs_status_ptr_t, ucp_stream_recv_nb,
+                 (ep, buffer, count, datatype, cb, flags),
+                 ucp_ep_h ep, void *buffer, size_t *count,
+                 ucp_datatype_t datatype, ucp_stream_recv_callback_t cb,
+                 unsigned flags)
+{
+    ucp_ep_ext_stream_t *ep_stream;
+    ucp_request_t       *req;
+    ucs_status_ptr_t    rdesc;
+    ucs_status_t        status;
+
+    if (UCP_DT_IS_GENERIC(datatype)) {
+        ucs_error("ucp_stream_recv_nb doesn't support generic datatype");
+        return UCS_STATUS_PTR(UCS_ERR_NOT_IMPLEMENTED);
+    }
+
+    UCP_THREAD_CS_ENTER_CONDITIONAL(&ep->worker->mt_lock);
+
+    req = ucp_stream_recv_request_get(ep->worker, buffer, *count, datatype, cb);
+    if (ucs_unlikely(req == NULL)) {
+        req = UCS_STATUS_PTR(UCS_ERR_NO_MEMORY);
+        goto out;
+    }
+
+    ep_stream = ep->ext.stream;
+
+    /* if there are already posted requests, need to queue this one */
+    if (!ucs_queue_is_empty(&ep_stream->reqs)) {
+        ucs_queue_push(&ep_stream->reqs, &req->recv.queue);
+        goto out;
+    }
+
+    /* OK, lets obtain all arrived data which matches the recv size */
+    while (req->recv.state.offset < req->recv.length) {
+        rdesc = ucp_stream_recv_data_nb_internal(ep_stream);
+        if (rdesc == NULL) {
+            /* No data any more. Enqueue empty request or complete if some
+             * data is filled */
+            if (req->recv.state.offset == 0) {
+                ucs_queue_push(&ep_stream->reqs, &req->recv.queue);
+            } else {
+                req = ucp_stream_recv_request_put(req, count, UCS_OK);
+            }
+            goto out;
+        }
+
+        status = ucp_stream_process_rdesc(rdesc, ep_stream, req);
+        if (ucs_unlikely(status != UCS_OK)) {
+            req = ucp_stream_recv_request_put(req, count, status);
+            goto out;
+        }
+    }
+
+    ucs_assert(req->recv.state.offset == req->recv.length);
+    req = ucp_stream_recv_request_put(req, count, UCS_OK);
+
+out:
+    UCP_THREAD_CS_EXIT_CONDITIONAL(&ep->worker->mt_lock);
+    return UCS_PTR_IS_PTR(req) ? (req + 1) : req;
```

Comment:
this branch is redundant. pls rearrange the return flow of the function to avoid. for example, can have 2 exit labels: "err" for errors and "out" for success
also, compiler can't optimize it because it has no knowledge about the value of the pointer pulled from mpool.